research/eccv2024/export_to_webdataset_and_crops.py

```diff
@@ -0,0 +1,215 @@
+                )
+
+                # Save the insect crop as image
+                insect_crop.save(os.path.join(export_dir, "insect_crops", str(binary_crop_count)+".jpg"))
+
+                # Export to webdataset for binary classification
+                if label_name!="Non-Moth": binary_class = "Moth"
+                else: binary_class = "Non-Moth"
+                sample_binary_annotation = {"label": binary_class, "region": region}
+                sample_binary = {
+                    "__key__": img_basename + "_" + str(binary_crop_count),
+                    "jpg": insect_crop,
+                    "json": sample_binary_annotation,
+                }
+                sink_binary.write(sample_binary)
+
+                # Save the binary annotation for the individual crop
+                label_binary[str(binary_crop_count)+".jpg"] = sample_binary_annotation
+
+                # Export to webdataset for finegrained classification, if moth crop
+                if label_rank != "NA":
+                    # If exists, get the synonym name and gbif keys for the species
+                    synonym = None
+                    speciesKey = None
+                    acceptedTaxonKey = None
+                    if label_rank == "SPECIES":
+                        synonym = get_synonym(sp_checklist, label_name)
+                        speciesKey, acceptedTaxonKey = get_gbif_keys(
+                            sp_key_map, sp_checklist, label_name
+                        )
+
+                    # Export to webdataset for fine-grained classification
+                    sample_fgrained_annotation = {
+                        "taxon_rank": label_rank,
+                        "label": label_name,
+                        "synonym": synonym,
+                        "speciesKey": speciesKey,
+                        "acceptedTaxonKey": acceptedTaxonKey,
+                        "region": region,
+                    }
+                    sample_fgrained = {
+                        "__key__": img_basename + "_" + str(fgrained_crop_count),
+                        "jpg": insect_crop,
+                        "json": sample_fgrained_annotation,
+                    }
+                    sink_fgrained.write(sample_fgrained)
+
+                    # Save the finegrained annotation for the individual crop
+                    label_fgrained[str(binary_crop_count)+".jpg"] = sample_fgrained_annotation
+                    fgrained_crop_count += 1
+
+                binary_crop_count += 1
+
+    sink_binary.close()
+    sink_fgrained.close()
+    with open(os.path.join(export_dir, "insect_crops", "binary_labels.json"), "w") as f:
+        json.dump(label_binary, f, indent=3)
+    with open(os.path.join(export_dir, "insect_crops", "fgrained_labels.json"), "w") as f:
+        json.dump(label_fgrained, f, indent=3)
+    print(f"The export of the AMI-Traps dataset to webdataset and individual crops is complete!", flush=True)
```

Comment:
This function would ideally need to be split up into distinct functions. As it is, this function both orchestrates the different steps, while also doing the work itself (processing the different steps).

Each step should ideally be it's own function, which would then be orchestrated by the `export_to_webdataset_and_crops` function, which would improve both the testability and the readability of this script/function.

Reply:
You can take a look at my PR here #15 where I did similar things.

Reply:
Fixed in ce73ced.